subsys/mpsl/clock_ctrl/mpsl_clock_ctrl.c

```diff
@@ -0,0 +1,299 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/clock_control/nrf_clock_control.h>
+
+#include <mpsl_clock.h>
+#include "mpsl_clock_ctrl.h"
+
+/* Variable shared for nrf and nrf2 clock control */
+static atomic_val_t m_hfclk_refcnt;
+
+/* Type use to get information about a clock request status */
+struct clock_onoff_state {
+	struct onoff_client cli;
+	atomic_t m_clk_ready;
+	atomic_val_t m_clk_refcnt;
+};
+
+static struct clock_onoff_state m_lfclk_state;
+
+/** @brief Wait for LFCLK to be ready.
+ *
+ * In case of nrf clock control driver the function does blcoking wait until the LFLCK is ready.
+ * In case of nrf2 clock control driver the function can time out in case there is no response
+ * from remote clock controller.
+ *
+ * @note For nRF54H SoC series waiting for LFCLK can't block the system work queue. In such case
+ *       the nrf2 clock control driver could return -TIMEDOUT due to lack of response from sysctrl.
+ *
+ * @retval 0 LFCLK is ready.
+ * @retval -NRF_EINVAL There were no LFCLK request.
+ * @retval -NRF_ETIMEDOUT LFCLK wait timed out.
+ */
+static int32_t m_lfclk_wait(void)
+{
+	int32_t err;
+	int32_t res;
+
+	if (m_lfclk_state.m_clk_ready == true) {
+		return 0;
+	}
+
+	/* Check if lfclk was already requested */
+	if (m_lfclk_state.m_clk_refcnt <= 0) {
+		return -NRF_EINVAL;
+	}
+
+	do {
+		err = sys_notify_fetch_result(&m_lfclk_state.cli.notify, &res);
+		if (!err && res) {
+			__ASSERT(false, "LFCLK could not be started: %d", res);
+			/* Translate -ERRTIMEDOUT to NRF counterpart */
+			return -NRF_ETIMEDOUT;
+		}
+		/* The function is executed in POST_KERNEL hence sleep is allowed */
+		k_msleep(1);
+	} while (err == -EAGAIN);
+
+	/* In case there is some other error. */
+	if (err < 0) {
+		return err;
+	}
+
+	atomic_set(&m_lfclk_state.m_clk_ready, true);
+
+	return 0;
+}
+
+#if defined(CONFIG_CLOCK_CONTROL_NRF)
+
+static void m_hfclk_request(void)
+{
+	/* The z_nrf_clock_bt_ctlr_hf_request doesn't count references to HFCLK,
+	 * it is caller responsibility handle requests and releases counting.
+	 */
+	if (atomic_inc(&m_hfclk_refcnt) > 0) {
+		return;
+	}
+
+	z_nrf_clock_bt_ctlr_hf_request();
+}
+
+static void m_hfclk_release(void)
+{
+	/* The z_nrf_clock_bt_ctlr_hf_request doesn't count references to HFCLK,
+	 * it is caller responsibility to do not release the clock if there is
+	 * other request pending.
+	 */
+	if (m_hfclk_refcnt < 1) {
+		return;
+	}
+
+	if (atomic_dec(&m_hfclk_refcnt) > 1) {
+		return;
+	}
+
+	z_nrf_clock_bt_ctlr_hf_release();
+}
+
+static bool m_hfclk_is_running(void)
+{
+	/* As of now assume the HFCLK is runnig after the request was put */
```

Comment:
Is this assumption valid?

Reply:
Well there is not much we can do about it here. The only thing that comes to my mind is use of GRTC->SYSCOUNTER and check if the call was done late enough to ensure the HFXO is ready. We do that internally, that way we could drop the check from withing the MPSL.

Reply:
It would be nice to add a comment here describing that this is not always true.
Maybe something like 
```suggestion
	/* As of now assume the HFCLK is running after the request was put.
	 * This puts the responsibility to the user to check if the time since last
	 * request is larger than the HFXO rampup time.
	 */
```

Another approach would be to disallow calling this API (`__ASSERT_NO_MSG(false) `) to make it clear that the result of this function call cannot be trusted. The user would anyways have check the time since the last request.

Reply:
I've changed the implementation for `nrf clock control` to use `nrfx_clock_is_running` to check if HFXO is ready. `nrf2 clock control` hasn't got this capability.